### PR TITLE
Feature/sdk v3.0/network layer/user data mapper

### PR DESCRIPTION
**Pull request by suriksarkisyan**

### CODE REVIEW

File: sdk/src/main/java/com/qonversion/android/sdk/internal/extensions/mapping/MappingExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.qonversion.android.sdk.internal.extensions.mapping
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/extensions/mapping/MappingExtensions.kt:1]: предлагаю с маленькой буквы назвать файл и просто extensions, тк он уже лежит в пакете mapping. А с маленькой, потому что в нём нет никаких классов, это утилитарный файл

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/ProductMapper.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.qonversion.android.sdk.internal.mappers
+
+import com.qonversion.android.sdk.public.Product
+
+class ProductMapper : Mapper<Product> {
+    override fun fromMap(data: Map<String, Any?>): Product? {
+        TODO("Not yet implemented")
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/ProductMapper.kt:7]: todo left here

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package com.qonversion.android.sdk.internal.mappers
+
+import com.qonversion.android.sdk.internal.extensions.mapping.getString
+import com.qonversion.android.sdk.internal.extensions.mapping.getDate
+import com.qonversion.android.sdk.internal.extensions.mapping.getMap
+import com.qonversion.android.sdk.public.Entitlement
+import com.qonversion.android.sdk.public.User
+import com.qonversion.android.sdk.public.UserPurchase
+
+class UserMapper internal constructor(
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt:10]: what is the need of internal constructor? We say that objects of this class can be created only in this package, but used anywhere. Is it the desired behavior?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    private val purchasesMapper: Mapper<UserPurchase>,
+    private val entitlementMapper: Mapper<Entitlement>
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt:12]: We can use concrete mapper classes here. No need to use interface with declared generic type, I guess. Discussable

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt [hunk 5]
```diff
@@ -2,8 +2,11 @@ package com.qonversion.android.sdk.internal.mappers
...
     @Throws(IllegalStateException::class)
-    fun toMap(value: T): Map<String, Any?>
+    fun toMap(value: T): Map<String, Any?> { return emptyMap() }
 
     @Throws(IllegalStateException::class)
-    fun fromMap(data: Map<String, Any?>): T
+    fun fromMap(data: Map<String, Any?>): T?
+
+    @Throws(IllegalStateException::class)
+    fun arrayFromMap(data: Map<*, *>): List<T> { return emptyList() }
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:11]: what is the use case of this method?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt [hunk 6]
```diff
@@ -2,8 +2,11 @@ package com.qonversion.android.sdk.internal.mappers
 
 interface Mapper<T> {
     @Throws(IllegalStateException::class)
-    fun toMap(value: T): Map<String, Any?>
+    fun toMap(value: T): Map<String, Any?> { return emptyMap() }
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:5]: I don't like that it has default implementation, cause we might forget to add custom one on some mappers but call it then, getting empty map always. And also don't this we should implement it in every mapper. Maybe we'd better divide this in two different interfaces?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserPurchaseMapper.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.qonversion.android.sdk.internal.mappers
+
+import com.qonversion.android.sdk.public.UserPurchase
+
+class UserPurchaseMapper : Mapper<UserPurchase> {
+    override fun fromMap(data: Map<String, Any?>): UserPurchase? {
+        TODO("Not yet implemented")
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserPurchaseMapper.kt:7]: todo

File: sdk/src/main/java/com/qonversion/android/sdk/public/PeriodType.kt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+package com.qonversion.android.sdk.public
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/public/PeriodType.kt:1]: I guess all the dto objects should be placed in the corresponding package inside `public`

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt (hunk 4 above)
Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt:12]: I think it's a bad idea. We should use the interface everywhere we can. Otherway we can't replace implementation w/o replacing declaration right here. Maybe it's not such a popular case when you need to replace mapper implementation, but we should have a standardized approach.

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt (hunk 5 above)
Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:11]:
```
{
    data: "some string",
    data2: [
            {}, {}, {}
           ]
}
```

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt (hunk 6 above)
Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:5]: For now, it's just a mock

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt (hunk 5 above)
Comment (SpertsyanKM) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:11]: And what is the expected result?

Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/Mapper.kt:11]: An array of objects (T)
Entitlements, for example

File: sdk/src/main/java/com/qonversion/android/sdk/public/Entitlement.kt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.qonversion.android.sdk.public
+
+import java.util.Date
+
+data class Entitlement(
+    val id: String,
+    val userID: String,
+    val active: Boolean,
+    val startDate: Date,
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/public/Entitlement.kt:9]: val startedDate: Date

File: sdk/src/main/java/com/qonversion/android/sdk/public/Subscription.kt [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.qonversion.android.sdk.public
+
+import java.util.Date
+
+data class Subscription(
+    val duration: ProductDuration,
+    val startDate: Date,
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/public/Subscription.kt:7]: val started
```suggestion
    val startedDate: Date,
```

File: sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        val createdDate = data.getDate("created")
+        val lastOnlineDate = data.getDate("last_online")
+
+        val entitlementsData = data.getMap("entitlements")
+        var entitlements: List<Entitlement> = emptyList()
+
+        val purchasesData = data.getMap("purchases")
+        var purchases: List<UserPurchase> = emptyList()
+
+        if (entitlementsData != null) {
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/mappers/UserMapper.kt:31]: Is there is a way to move null check into the `arrayFromMap` func? 
`arrayFromMap` will return empty array for null value

**Review by Maria-Bordunova — APPROVED** [review]

**Review by SpertsyanKM — APPROVED** [review]

**Pull request merged**